src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs

```diff
@@ -294,4 +348,54 @@ public static ObjectPool<ExpressionVariableFinder> CreatePool()
         }
         #endregion
     }
+
+    internal class ExpressionFieldFinder : ExpressionVariableFinder<Symbol>
+    {
+        private SourceMemberContainerTypeSymbol _containingType;
+        private DeclarationModifiers _modifiers;
+        private FieldSymbol _containingFieldOpt;
+
+        internal static void FindExpressionVariables(
+            ArrayBuilder<Symbol> builder,
+            CSharpSyntaxNode node,
+            SourceMemberContainerTypeSymbol containingType,
+            DeclarationModifiers modifiers,
+            FieldSymbol containingFieldOpt)
+        {
+            if (node == null)
+            {
+                return;
+            }
+
+            var finder = s_poolInstance.Allocate();
+            finder._containingType = containingType;
+            finder._modifiers = modifiers;
+            finder._containingFieldOpt = containingFieldOpt;
+
+            finder.FindExpressionVariables(builder, node);
+
+            finder._containingType = null;
+            finder._modifiers = DeclarationModifiers.None;
+            finder._containingFieldOpt = null;
+            s_poolInstance.Free(finder);
+        }
+
+        protected override Symbol MakePatternVariable(DeclarationPatternSyntax node, SyntaxNode nodeToBind)
+        {
+            throw ExceptionUtilities.Unreachable;
```

Comment:
Throwing an exception doesn't seem like the right way to handle it. We've generally handled these things using diagnostics. We reserve throwing an unreachable exception for code that we believe is unreachable.

---
In reply to: [79079667](https://github.com/dotnet/roslyn/pull/13818#discussion_r79079667) [](ancestors = 79079667)